games/maze/src/maze.rs

```diff
@@ -57,24 +66,120 @@ pub enum Direction {
-            map: vec![Tile::blocked(); size * size],
+            map: random_map,
         };
 
+        maze.reveal_around_player();
+        maze
+    }
+
+    fn generate_random_map(size: usize) -> Vec<Tile> {
+        fn find(
+            size: usize,
+            map: &[MazeGenerationTile],
+            (p, q): (Position, Position),
+        ) -> (Position, Position) {
+            let (cell_p, cell_q) = (map[size * p.y + p.x].link, map[size * q.y + q.x].link);
+
+            if p != cell_p || q != cell_q {
+                find(size, map, (cell_p, cell_q))
+            } else {
+                (cell_p, cell_q)
+            }
+        }
+
+        if size & 1 == 0 {
+            panic!("Random maze only allows odd numbers")
+        };
+
+        let mut gen_map = Vec::with_capacity(size * size);
+
         for i in 0..size {
-            maze.set(i, 0, Tile::open());
-            maze.set(size - 1, i, Tile::open());
+            for j in 0..size {
+                let pos = Position { x: j, y: i };
+
+                match (j & 1 == 0, i & 1 == 0) {
+                    (true, true) => gen_map.push(MazeGenerationTile {
+                        position: pos,
+                        link: pos,
+                        tile_type: Some(TileType::Open),
+                    }),
+                    (false, false) => gen_map.push(MazeGenerationTile {
+                        position: pos,
+                        link: pos,
+                        tile_type: Some(TileType::Blocked),
+                    }),
+                    (false, true) | (true, false) => gen_map.push(MazeGenerationTile {
+                        position: pos,
+                        link: pos,
+                        tile_type: None,
+                    }),
+                }
+            }
         }
 
-        maze.reveal_around_player();
+        let neither_map = gen_map.to_vec();
+        let mut neither_map = neither_map
+            .iter()
+            .filter(|x| match x.tile_type {
+                None => true,
+                _ => false,
+            })
+            .collect::<Vec<_>>();
+
+        neither_map.shuffle(&mut thread_rng());
+
+        for i in neither_map {
+            let find_map = |x| find(size, &gen_map, x);
+            let pos = i.position;
+
+            let (p, q) = if pos.y & 1 == 0 {
+                find_map((
+                    Position {
+                        x: pos.x + 1,
+                        y: pos.y,
+                    },
+                    Position {
+                        x: pos.x - 1,
+                        y: pos.y,
+                    },
+                ))
+            } else {
+                find_map((
+                    Position {
+                        x: pos.x,
+                        y: pos.y - 1,
+                    },
+                    Position {
+                        x: pos.x,
+                        y: pos.y + 1,
+                    },
+                ))
+            };
```

Comment:
The `find_map` closure does not need to be defined inside the loop. Or at all if you invert this if and inline the closure:

```rust
    let pos = i.position;

    let (p, q) = find(
        size,
        &gen_map,
        if pos.y & 1 == 0 {
            (
                Position {
                    x: pos.x + 1,
                    y: pos.y,
                },
                Position {
                    x: pos.x - 1,
                    y: pos.y,
                },
            )
        } else {
            (
                Position {
                    x: pos.x,
                    y: pos.y - 1,
                },
                Position {
                    x: pos.x,
                    y: pos.y + 1,
                },
            )
        },
    );
```